utils/src/priority_set.rs

```diff
@@ -0,0 +1,180 @@
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::hash::Hash;
+
+/// An entry in the `PrioritySet`.
+#[derive(Eq, PartialEq)]
+struct Entry<I: Ord + Hash + Clone, P: Ord + Copy> {
+    item: I,
+    priority: P,
+}
+
+impl<I: Ord + Hash + Clone, P: Ord + Copy> Ord for Entry<I, P> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.priority.cmp(&other.priority) {
+            Ordering::Equal => self.item.cmp(&other.item),
+            other => other,
+        }
+    }
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Copy> PartialOrd for Entry<I, V> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A set that offers efficient iteration over
+/// its elements in priority-ascending order.
+pub struct PrioritySet<I: Ord + Hash + Clone, P: Ord + Copy> {
+    entries: BTreeSet<Entry<I, P>>,
+    keys: HashMap<I, P>,
+}
+
+impl<I: Ord + Hash + Clone, P: Ord + Copy> PrioritySet<I, P> {
+    /// Create a new `PrioritySet`.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            entries: BTreeSet::new(),
+            keys: HashMap::new(),
+        }
+    }
+
+    /// Insert an item with a priority, overwriting the previous priority if it exists.
+    pub fn put(&mut self, item: I, priority: P) {
+        // Remove old entry, if it exists
+        let entry = if let Some(old_priority) = self.keys.remove(&item) {
+            // Remove the item from the old priority's set
+            let mut old_entry = Entry {
+                item: item.clone(),
+                priority: old_priority,
+            };
+            self.entries.remove(&old_entry);
+
+            // We reuse the entry to avoid another item clone
+            old_entry.priority = priority;
+            old_entry
+        } else {
+            Entry { item, priority }
+        };
+
+        // Insert the entry
+        self.keys.insert(entry.item.clone(), entry.priority);
+        self.entries.insert(entry);
+    }
+
+    /// Get the current priority of an item.
+    pub fn get(&self, item: &I) -> Option<P> {
+        self.keys.get(item).cloned()
+    }
+
+    /// Remove an item from the set.
+    pub fn remove(&mut self, item: &I) {
+        let Some(entry) = self.keys.remove(item).map(|priority| Entry {
+            item: item.clone(),
+            priority,
+        }) else {
+            return;
```

Comment:
nit: looks like this line is untested, consider a simple test to remove an item that doesn't exist